csharp/extractor/Semmle.Extraction/EscapingTextWriter.cs

```diff
@@ -0,0 +1,340 @@
+using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Semmle.Extraction
+{
+    /// <summary>
+    /// A `TextWriter` object that wraps another `TextWriter` object, and which
+    /// HTML escapes the characters `&`, `{`, `}`, `"`, `@`, and `#`, before
+    /// writing to the underlying object.
+    /// </summary>
+    public class EscapingTextWriter : TextWriter
+    {
+        private readonly TextWriter wrapped;
+        private readonly bool disposeUnderlying;
+
+        public EscapingTextWriter(TextWriter wrapped, bool disposeUnderlying = false)
+        {
+            this.wrapped = wrapped;
+            this.disposeUnderlying = disposeUnderlying;
+        }
+
+        /// <summary>
+        /// Creates a new instance with a new underlying `StringWriter` object. The
+        /// underlying object is disposed of when this object is.
+        /// </summary>
+        public EscapingTextWriter() : this(new StringWriter(), true) { }
+
+        public EscapingTextWriter(IFormatProvider? formatProvider) : base(formatProvider)
+            => throw new NotImplementedException();
+
+        private void WriteEscaped(char c)
+        {
+            switch (c)
+            {
+                case '&':
+                    wrapped.Write("&amp;");
+                    break;
+                case '{':
+                    wrapped.Write("&lbrace;");
+                    break;
+                case '}':
+                    wrapped.Write("&rbrace;");
+                    break;
+                case '"':
+                    wrapped.Write("&quot;");
+                    break;
+                case '@':
+                    wrapped.Write("&commat;");
+                    break;
+                case '#':
+                    wrapped.Write("&num;");
+                    break;
+                default:
+                    wrapped.Write(c);
+                    break;
+            };
+        }
+
+        public void WriteSubId(IEntity entity)
+        {
+            if (entity is null)
+            {
+                wrapped.Write("<null>");
+                return;
+            }
+
+            WriteUnescaped('{');
+            wrapped.WriteLabel(entity);
+            WriteUnescaped('}');
+        }
+
+        public void WriteUnescaped(char c)
+            => wrapped.Write(c);
+
+        public void WriteUnescaped(string s)
+            => wrapped.Write(s);
+
+        #region overrides
+
+        public override Encoding Encoding => wrapped.Encoding;
+
+        public override IFormatProvider FormatProvider => wrapped.FormatProvider;
+
+        public override string NewLine { get => wrapped.NewLine; }
+
+        public override void Close()
+            => throw new NotImplementedException();
+
+        public override ValueTask DisposeAsync()
+            => throw new NotImplementedException();
```

Comment:
Why aren't these calling the `wrapped`?

Reply:
I didn't bother to implement those because we are not using them.